File: android/app/src/main/kotlin/de/yamt/app/homewidget/QuickEatBar.kt
package de.yamt.app.homewidget

import android.content.Context
import android.net.Uri
import androidx.compose.runtime.Composable
import androidx.compose.ui.unit.dp
import androidx.glance.ColorFilter
import androidx.glance.GlanceModifier
import androidx.glance.Image
import androidx.glance.ImageProvider
import androidx.glance.action.Action
import androidx.glance.action.clickable
import androidx.glance.appwidget.cornerRadius
import androidx.glance.background
import androidx.glance.layout.Alignment
import androidx.glance.layout.Box
import androidx.glance.layout.Row
import androidx.glance.layout.Spacer
import androidx.glance.layout.fillMaxWidth
import androidx.glance.layout.height
import androidx.glance.layout.size
import androidx.glance.layout.width
import de.yamt.app.MainActivity
import de.yamt.app.R
import es.antonborri.home_widget.actionStartActivity

/**
 * Compact DiaryQuickEatBar: glyph-only buttons, height 44, radius md (14),
 * surfaceContainerLow, gap xs (8). Search, AI and barcode open the product
 * search hub (HomeWidgetClickAction parses the uri).
 */
@Composable
internal fun QuickEatBar(context: Context) {
  val buttons =
      listOf(
          Triple(R.drawable.ic_widget_search, R.string.home_widget_action_search, "quick-add?intent=search"),
          Triple(R.drawable.ic_widget_ai, R.string.home_widget_action_ai, "quick-add?intent=ai"),
          Triple(R.drawable.ic_widget_barcode, R.string.home_widget_action_barcode, "quick-add?intent=barcode"),
      )
  Row(modifier = GlanceModifier.fillMaxWidth()) {
    buttons.forEachIndexed { index, (icon, description, path) ->
      if (index > 0) {
        Spacer(modifier = GlanceModifier.width(8.dp))
      }
      Box(
          modifier =
              GlanceModifier.defaultWeight()
                  .height(44.dp)
                  .background(HomeWidgetColors.surfaceContainerLow)
                  .cornerRadius(14.dp)
                  .clickable(openAppAction(context, "homewidget://$path")),
          contentAlignment = Alignment.Center,
      ) {
        Image(
            provider = ImageProvider(icon),
            contentDescription = context.getString(description),
            colorFilter = ColorFilter.tint(HomeWidgetColors.onSurface),
            modifier = GlanceModifier.size(20.dp),
        )
      }
    }
  }
}

/**
 * Opens [MainActivity] with [uri] as the launch data home_widget hands to
 * Flutter. Always pass a uri: a null Intent.data collides with Glance's own
 * glance_actions://CALLBACK action dispatch. HomeWidgetClickAction ignores
 * any uri that isn't a quick-add action, so those just open the app.
 */
internal fun openAppAction(context: Context, uri: String): Action =
    actionStartActivity<MainActivity>(context, Uri.parse(uri))

File: ios/HomeDiaryWidget/QuickEatBar.swift
import SwiftUI

/// Compact DiaryQuickEatBar: glyph-only buttons, height 44, radius md (14),
/// surfaceContainerLow, gap xs (8). Search, AI and barcode open the product
/// search hub (HomeWidgetClickAction parses the uri). Same buttons as the
/// Android widget's QuickEatBar.kt.
struct QuickEatBar: View {
  private struct QuickEatButton {
    let symbol: String
    let descriptionKey: String
    let path: String
  }

  private let buttons = [
    QuickEatButton(symbol: "magnifyingglass", descriptionKey: "home_widget_action_search", path: "quick-add?intent=search"),
    QuickEatButton(symbol: "sparkles", descriptionKey: "home_widget_action_ai", path: "quick-add?intent=ai"),
    QuickEatButton(symbol: "barcode.viewfinder", descriptionKey: "home_widget_action_barcode", path: "quick-add?intent=barcode"),
  ]

  var body: some View {
    HStack(spacing: 8) {
      ForEach(buttons, id: \.path) { button in
        Link(destination: homeWidgetUrl(button.path)) {
          Image(systemName: button.symbol)
            .font(.system(size: 18, weight: .medium))
            .foregroundColor(HomeWidgetColors.onSurface)
            .frame(maxWidth: .infinity)
            .frame(height: 44)
            .background(HomeWidgetColors.surfaceContainerLow)
            .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
            .accessibilityLabel(localized(button.descriptionKey))
        }
      }
    }
  }
}

/// `homewidget://<path>` plus the `homeWidget` query item the home_widget
/// plugin needs to hand the URL to Flutter. HomeWidgetClickAction ignores
/// any uri that isn't a quick-add action, so those just open the app.
func homeWidgetUrl(_ path: String) -> URL {
  let separator = path.contains("?") ? "&" : "?"
  return URL(string: "homewidget://\(path)\(separator)homeWidget")!
}
